code/datums/components/window_smoothing.dm

```diff
@@ -0,0 +1,134 @@
+	var/atom/parent = src.parent
+	var/junction = new_junction
+	if(isnull(junction))
+		junction = parent.smoothing_junction
+
+	var/turf/our_turf = get_turf(parent)
+	var/offset = GET_Z_PLANE_OFFSET(our_turf.z)
+	// Alright. First we're gonna decide how to display our lower section
+	// So we'll start with "do we have a wall down there"
+	// we check to the south east and west too because action hates me
+	// OF NOTE:
+	// what we are checking here is if there's a wall below, OR if there's a wall to the south(east|west) and
+	// a SMOOTHING TARGET to the south.
+	// This ensures we only use the alt display if there's a wall there, AND we're smoothing into it
+	var/wall_below = (isclosedturf(get_step(our_turf, SOUTH)) && (!ignored_turf || !istype(get_step(our_turf, SOUTH), ignored_turf))) || \
+		(junction & (SOUTH) && isclosedturf(get_step(our_turf, SOUTHEAST)) && (!ignored_turf || !istype(get_step(our_turf, SOUTHWEST), ignored_turf))) || \
+		(junction & (SOUTH) && isclosedturf(get_step(our_turf, SOUTHWEST)) && (!ignored_turf || !istype(get_step(our_turf, SOUTHWEST), ignored_turf)))
+	// If there's a wall below us, we render different
+	our_appearances += get_window_appearance(offset, icon_path, junction, "lower", wall_below)
+
+	if(paired_turf)
+		stack_trace("We tried to generate a connection to a turf while one already exists")
+	// Now we'll check above
+	// If there's nothin, we'll use a frill. Otherwise we won't
+	paired_turf = get_step(our_turf, NORTH)
+	// wallening TODO: ok this shit ain't working. Instead of trying to track walls like this, we should attach an effect object to the turf of our window that smooths for walls, and pass that into window logic
+
+	// We only display an above state if there's a wall, OR if we're smoothing with nothing up there
+	if(isclosedturf(paired_turf) && (!ignored_turf || !istype(paired_turf, ignored_turf)))
+		our_appearances += get_window_appearance(offset, icon_path, junction, "upper", TRUE, pixel_y = 32)
+		UnregisterSignal(paired_turf, COMSIG_QDELETING)
+	else if(!(NORTH & junction))
+		// Draw to the turf above you so this can be seen without seeing the window's turf. Oh and draw this as a frill
+		// We use the parent's pixel y as a part of this to ensure everything lines up proper when the parent is all shifted around
+		appearance_above = get_window_appearance(offset, icon_path, junction, "upper", FALSE, pixel_y = parent.pixel_y, plane = FRILL_PLANE)
+		paired_turf.add_overlay(appearance_above)
+		RegisterSignal(paired_turf, COMSIG_QDELETING, PROC_REF(tied_turf_deleted), override = TRUE) // Override because this could be called multiple times
+	else
+		UnregisterSignal(paired_turf, COMSIG_QDELETING)
+
+	parent.add_overlay(our_appearances)
+
+/datum/component/window_smoothing/proc/remove_smoothing()
+	var/atom/atom_parent = parent
+	atom_parent.cut_overlay(our_appearances)
+	our_appearances.Cut()
+
+	paired_turf.cut_overlay(appearance_above)
```

Comment:
```suggestion
	paired_turf.cut_overlay(appearance_above)
	appearance_above = null
```
`add_smoothing` will generate a new `appearance_above`, this is to prevent hard dels.

Reply:
should not be an issue, appearance_above is a mutable appearance and we do not garbage collect those

Reply:
Oh ok.